runelite-client/src/test/java/net/runelite/client/plugins/cannon/CannonPluginTest.java

```diff
@@ -171,6 +179,145 @@ public void unloadCannon()
 		assertEquals(0, plugin.getCballsLeft());
 	}
 
+	@Test
+	public void cannonDestroyedSetupSameSession()
+	{
+		setUpCannon();
+		plugin.onChatMessage(loadCannonballs(5));
+		plugin.onChatMessage(CANNON_DESTROYED);
+		setUpCannon();
+
+		assertTrue(plugin.isCannonPlaced());
+
+		plugin.onChatMessage(loadCannonballs(25));
+
+		assertEquals(25, plugin.getCballsLeft());
+	}
+
+	@Test
+	public void cannonDestroyedSetupNormalCannonballsDifferentSession()
+	{
+		setUpCannon();
+		plugin.onChatMessage(loadCannonballs(5));
+		plugin.onChatMessage(CANNON_DESTROYED);
+
+		plugin.shutDown();
+		plugin.startUp();
+
+		setUpCannon();
+
+		assertTrue(plugin.isCannonPlaced());
+
+		final ItemContainer inventory = mock(ItemContainer.class);
+		final Item[] inventoryContentsBefore = { new Item(ItemID.CANNONBALL, 100), new Item(ItemID.GRANITE_CANNONBALL, 100) };
+		when(inventory.getItems()).thenReturn(inventoryContentsBefore);
+		when(inventory.contains(ItemID.CANNONBALL)).thenReturn(true);
+		lenient().when(inventory.contains(ItemID.GRANITE_CANNONBALL)).thenReturn(true);
+		when(client.getItemContainer(InventoryID.INVENTORY)).thenReturn(inventory);
+
+		plugin.onChatMessage(loadCannonballs(25));
+
+		final Item[] inventoryContentsAfter = { new Item(ItemID.CANNONBALL, 75), new Item(ItemID.GRANITE_CANNONBALL, 100) };
```

Comment:
This doesn't really follow in game behavior since it will always load granite cannonballs if they exist in your inventory, no matter what, but for the purposes of testing code paths this could be fine.